### ISSUE TITLE
Allow implementing Protocol variables as properties

### ISSUE THREAD

**Issue by williamw4096** [microsoft/pyright#7295]
An abstract protocol variable can only currently be implemented by defining a variable in the implementing class. Ideally implementing classes would be given much more flexibility  by allowing a property of the same name and type to do the job of implementing the protocol variable, since this still exposes the exact same API.

Code example:

```py
from typing import Protocol

class HasX(Protocol):
    x: int

class PropertyX:
    @property
    def x(self) -> int:
        return 1

    @x.setter
    def x(self, _) -> None:
        pass

has_x: HasX = PropertyX()
```

The last line would ideally type check, but currently gives 
```
"PropertyX is incompatible with protocol "HasX"
    "x" is invariant because it is mutable
    "x" is an incompatible type
        "property" is incompatible with "int"
```

Although I can't find any documentation to verify that this proposed behaviour is the expected behaviour of protocol variables, mypy does implement this feature.

**Comment by erictraut** [comment]
The typing spec is not clear on this issue (whether implementations of protocols can implement an abstract variable as a property), so the decision is currently left up to individual type checkers. If you would like to argue in favor of clarifying this in the typing spec, please consider starting a discussion in the [Python typing forum](https://discuss.python.org/c/typing/32).

My personal take (which is reflected in the current implementation) is that this should not be allowed. There are semantic differences between variables and descriptor objects (including properties). These differences include:
1. Variables are deletable, but unless you define a deleter for your property, it is not
2. Variables can be accessed from a class object; when you access a descriptor object from the class object, you'll receive the descriptor object, not the variable it is abstracting

For these reasons, I'm not inclined to make this change unless/until the typing community agrees that this should become the official behavior in the typing spec.

**Comment by shartzog** [comment]
@erictraut Any advice on the _proper_ way to allow this behavior? There doesn't appear to be any mechanism for allowing this kind of behavior at the moment. I'm inclined to agree with the OP. Protocols are about structure, and I'd argue that a gettable/settable property and an *instance* attribute are structurally equivalent. Nonetheless, I agree that the PEP is ambiguous, and this implementation detail is firmly in your court. Just adding my two cents in favor of support. ;)

**Comment by erictraut** [comment]
There was a recent discussion on this topic in the [Python typing forum](https://discuss.python.org/t/typing-spec-gap-about-properties-in-protocols/87243/4).
